### ISSUE TITLE
[IMPROVE] Forecasting models should allow different error measures as validation criteria

### ISSUE THREAD

**Issue by umitkaanusta** [OpenBB-finance/OpenBBTerminal#4136]
**Feature:** In the forecasting models, the best model is chosen based on MAPE during validation, allowing no other error metric to be used instead of MAPE. We can easily allow different error measures in the process.
**Why:** MAPE is not always the optimal error criterion. E.g. there is WAPE (Weighted Absolute Percentage Error) if one would like to take a weighted average of percent errors, weights being the magnitude of actuals. This is useful when forecasting, say, net profit and you'd like to have your model prioritize correctly predicting highly profitable companies.

**How to improve:**
Taking *theta* implementation as an example, the current state uses MAPE as the following:
```python3
        try:
            model.fit(train)
            pred_theta = model.predict(len(val))
            res = mape(val, pred_theta)
            if res < best_mape:
                best_mape = res
                best_theta = theta
        except Exception as e:  # noqa
            error = str(e)

```
This can be expanded in 2 ways:
i) Implementing different time-series accuracy metrics like Symmetric MAPE, WAPE, AIC (Akaike Info. Criterion) etc. and users can choose through a parameter like `metric`.
ii) Allowing users to define a custom accuracy metric with a structure like `func(actual, pred) -> score`, and accepting such functions through the parameter `metric` as well.

**A hypothetically improved version:**
The above code snippet would change as
```python3
        try:
            model.fit(train)
            pred_theta = model.predict(len(val))
            res = calculate_error(val, pred_theta, metric)
            if improves_on_best(res, best_res, metric):
                best_res = res
                best_theta = theta
        except Exception as e:  # noqa
            error = str(e)

```
The function `calculate_error` computes the given metric. If the metric is readily implemented, it fetches the related metric and returns the result. If not, it calls the function as `func(val, pred_theta)`. In the latter case, the function can be checked to see if it has the appropriate signature / other properties.

The function `improves_on_best` outputs a boolean, using `metric`'s `smaller_better` attribute.
Each metric is implemented in the module `/forecast/metrics.py`. Each metric has an attribute denoting if smaller is better (if not, bigger is better), and a function to calculate it.

**Comment by jmaslek** [comment]
#4178 closed this
